pkg/util/steps/runner.go

```diff
@@ -0,0 +1,186 @@
+// Name returns the "friendly name" of the function that the Step will execute.
+type Step interface {
+	run(ctx context.Context, runner *Runner) error
+	Name() string
+}
+
+type ActionFunction func(context.Context) error
+
+type actionStep struct {
+	f ActionFunction
+}
+
+// run executes the action function. If the action function returns an error, it
+// asks Runner.handler.ShouldRetryStep whether it should be rerun. If
+// ShouldRetryStep returns true, it will be reran without resetting the timeout.
+// If ShouldRetryStep returns false, the error will be returned.
+func (s actionStep) run(ctx context.Context, runner *Runner) error {
+	var (
+		shouldRerun bool
+		err         error
+	)
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, runner.stepTimeout)
+	defer cancel()
+
+	wait.PollImmediateUntil(runner.pollInterval, func() (bool, error) {
+		err = s.f(timeoutCtx)
+
+		if err != nil {
+			shouldRerun, err = runner.handler.ShouldRetryStep(timeoutCtx, err, s.f)
+		} else {
+			shouldRerun = false
+		}
+
+		return !shouldRerun, err
+	}, timeoutCtx.Done())
+
+	return err
+}
+
+func (s actionStep) Name() string {
+	return log.FriendlyName(s.f)
+}
+
+// Action returns a Step suitable for performing an action.
+//
+// The Action will execute f when the step is reached, timing out with a failure
+// when more time than Runner.stepTimeout has elapsed. If f returns an error,
+// the Runner will call StepResultHandler.ShouldRetryStep to check whether it
+// should be rerun (e.g. due to transient errors/unrefreshed credentials). If it
+// is not to be rerun, the Action will fail and no subsequent Steps will be run.
+func Action(f ActionFunction) actionStep {
+	return actionStep{f}
+}
+
+type ConditionFunction func(context.Context) (bool, error)
+
+type conditionStep struct {
+	f       ConditionFunction
+	timeout time.Duration
+}
+
+// Condition returns a Step suitable for checking whether subsequent Steps can
+// be executed.
+//
+// The Condition will execute f repeatedly (every Runner.pollInterval), timing
+// out with a failure when more time than the provided timeout has elapsed
+// without f returning (true, nil). If f returns err, the Condition will fail.
+// If the Condition times out or fails, no subsequent Steps will be run.
+func Condition(f ConditionFunction, timeout time.Duration) conditionStep {
+	return conditionStep{
+		f:       f,
+		timeout: timeout,
+	}
+}
+
+// run checks whether the condition function returns true every
+// runner.pollInterval. If the condition function returns an error, it will
+// instantly fail.
+func (c conditionStep) run(ctx context.Context, runner *Runner) error {
+	var (
+		err  error
+		done bool
+	)
+	timeoutCtx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+
+	// Run the function every runner.pollInterval until it returns true or times
+	// out.
+	wait.PollImmediateUntil(runner.pollInterval, func() (bool, error) {
+		done, err = c.f(timeoutCtx)
+		return done, err
+	}, timeoutCtx.Done())
```

Comment:
One of them didn't (https://github.com/Azure/ARO-RP/blob/318bd55a156f117f3d61528fb30f6f9de1a0e11f/pkg/install/install.go#L242-L252) , although I've now refactored it so that they both do.